### PR TITLE
Write snapshot explicity after truncate backend

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: Ifa11b36bebcd7e702327e79d51ec68fe1fd2eebb

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java [hunk 1]
```diff
@@ -182,9 +195,14 @@ public void onApply(Iterator iter) {
 
     private Object applyCommand(StoreType type, StoreAction action,
                                 BytesBuffer buffer) {
-        BackendStore store = this.store(type);
-        BiConsumer<BackendStore, BytesBuffer> func = this.funcs.get(action);
-        func.accept(store, buffer);
+        if (type != StoreType.ALL) {
+            BackendStore store = this.store(type);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:199]: set store=null if type == StoreType.ALL

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 2]
```diff
@@ -98,6 +98,19 @@ public void shutdown() {
...
+    public void snapshot() {
+        if (!this.context.useSnapshot()) {
+            return;
+        }
+        RaftClosure<?> future = new RaftClosure<>();
+        try {
+            this.node().snapshot(future);
+            future.waitFinished();
+        } catch (Throwable e) {
+            throw new BackendException("Failed to genearet snapshot", e);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:110]: typo: genearet

**Review by javeme — APPROVED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java [hunk 3]
```diff
@@ -188,6 +189,20 @@ public void initSystemInfo(HugeGraph graph) {
         LOG.debug("Graph '{}' system info has been initialized", this.graph());
     }
 
+    @Override
+    public void writeSnapshot() {
+        StoreCommand command = new StoreCommand(StoreType.ALL,
+                                                StoreAction.SNAPSHOT, null);
+        StoreClosure closure = new StoreClosure(command);
+        this.context.node().submitAndWait(command, closure);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java:197]: set submitAndWait protected

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java [hunk 4]
```diff
@@ -123,8 +127,8 @@ private void updateCacheIfNeeded(BackendMutation mutation) {
     }
 
     private void register(StoreAction action,
-                          BiConsumer<BackendStore, BytesBuffer> func) {
-        this.funcs.put(action, func);
+                          BiConsumer<BackendStore, BytesBuffer> biFunc) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:130]: keep func

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/rpc/RaftRequests.java [hunk 5]
```diff
@@ -3,14 +3,13 @@
 
 package com.baidu.hugegraph.backend.store.raft.rpc;
 
-@SuppressWarnings("unused")
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/rpc/RaftRequests.java:6]: keep @SuppressWarnings("unused")

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java (hunk 3 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java:197]: StoreCommandProcessor used it

**Review by javeme — APPROVED** [review]

**Review by zhoney — APPROVED** [review]

**Pull request merged**